confidential-data-hub/kms/src/plugins/ehsm/client.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2023 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use ehsm_client::{client::EHSMClient, api::KMS};
+
+use async_trait::async_trait;
+use base64::engine::general_purpose::STANDARD;
+use base64::Engine;
+use const_format::concatcp;
+use serde_json::Value;
+use tokio::fs;
+
+use crate::plugins::_IN_GUEST_DEFAULT_KEY_PATH;
+use crate::{Annotations, Decrypter, Encrypter, ProviderSettings};
+use crate::{Error, Result};
+
+use super::annotations::EhsmProviderSettings;
+use super::credential::Credential;
+
+pub struct EhsmKmsClient {
+    client: EHSMClient,
+}
+
+const EHSM_IN_GUEST_DEFAULT_KEY_PATH: &str = concatcp!(_IN_GUEST_DEFAULT_KEY_PATH, "/ehsm");
+
+impl EhsmKmsClient {
+    pub fn new(
+        app_id: &str,
+        api_key: &str,
+        endpoint: &str,
+    ) -> Result<Self> {
+        Ok(Self {
+            client: EHSMClient {
+                base_url: endpoint.to_owned(),
+                appid: app_id.to_owned(),
+                apikey: api_key.to_owned(),
+            }
+        })
+    }
+
+    /// build client with parameters that have been exported to environment.
+    pub fn new_from_env() -> Result<Self> {
+
+        Ok(Self {
+            client: EHSMClient::new(),
+        })
+    }
+
+    /// This new function is used by a in-pod client. The side-effect is to read the
+    /// [`EHSM_IN_GUEST_DEFAULT_KEY_PATH`] which is the by default path where the credential
+    /// to access kms is saved.
+    pub async fn from_provider_settings(provider_settings: &ProviderSettings) -> Result<Self> {
+        let provider_settings: EhsmProviderSettings =
+            serde_json::from_value(Value::Object(provider_settings.clone())).map_err(|e| {
+                Error::EhsmKmsError(format!("parse provider setting failed: {e}"))
+            })?;
+
+        let credential_path = format!(
+            "{EHSM_IN_GUEST_DEFAULT_KEY_PATH}/credential_{}.json",
+            provider_settings.app_id
```

Comment:
About the way to read credential. Currently we follow the steps:
1. CDH get credential from the KBS
2. CDH put the credential under `/run/..`
3. CDH reads the credential under `/run/..`

@fitzthum once mentioned that this will expose the plaintext of credentials to the guest and suggest that the credentials be kept in the memory of the CDH. I think it is a good way, and the steps could be
1. CDH get credential from the KBS
2. CDH sets its own process's env the credential with specific keys
3. Concrete KMS plugin reads the credential from the env

Which I'd like to put another PR and let's talk more about that then.

Reply:
Glad to hear that. eHSM-KMS itself supports reading the credential from environment variables, which will make the transition easy to implement.